### PR TITLE
interpret: simplify SIMD type handling

### PR DESCRIPTION
**Pull request by RalfJung**
This is possible as a follow-up to https://github.com/rust-lang/rust/pull/129403

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/place.rs [hunk 1]
```diff
@@ -391,7 +393,16 @@ where
         layout: TyAndLayout<'tcx>,
     ) -> MPlaceTy<'tcx, M::Provenance> {
         assert!(layout.is_sized());
-        self.ptr_with_meta_to_mplace(ptr, MemPlaceMeta::None, layout)
+        self.ptr_with_meta_to_mplace(ptr, MemPlaceMeta::None, layout, /*unaligned*/ false)
+    }
+
+    pub fn ptr_to_mplace_unaligned(
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/place.rs:399]: @tiif this method should also come handy in https://github.com/rust-lang/miri/pull/3852 to deal with the alignment problem in eventfd

File: src/tools/miri/src/shims/x86/mod.rs [hunk 2]
```diff
@@ -1000,23 +1000,26 @@ fn mask_store<'tcx>(
...
         let value = this.project_index(&value, i)?;
 
         if this.read_scalar(&mask)?.to_uint(mask_item_size)? >> high_bit_offset != 0 {
+            // *Bon-inbounds* pointer arithmetic to compute the destination.
             let ptr = ptr.wrapping_offset(value.layout.size * i, &this.tcx);
-            // Unaligned copy, which is what we want.
-            this.mem_copy(value.ptr(), ptr, value.layout.size, /*nonoverlapping*/ true)?;
+            // Deref the pointer *unaligned*, and do the copy.
+            let dest = this.ptr_to_mplace_unaligned(ptr, value.layout);
+            this.copy_op(&value, &dest)?;
```
Comment (RalfJung) [inline src/tools/miri/src/shims/x86/mod.rs:1022]: I had to adjust the logic a bit here because `operand_to_simd` used to return an `MPlace` but `project_to_simd` preserves the original type.

File: compiler/rustc_const_eval/src/interpret/place.rs (hunk 1 above)
Comment (tiif) [reply compiler/rustc_const_eval/src/interpret/place.rs:399]: Yup, thanks!

File: compiler/rustc_const_eval/src/interpret/util.rs [hunk 3]
```diff
@@ -103,5 +103,5 @@ pub(crate) fn create_static_alloc<'tcx>(
     assert_eq!(ecx.machine.static_root_ids, None);
     ecx.machine.static_root_ids = Some((alloc_id, static_def_id));
     assert!(ecx.memory.alloc_map.insert(alloc_id, (MemoryKind::Stack, alloc)).is_none());
-    Ok(ecx.ptr_with_meta_to_mplace(Pointer::from(alloc_id).into(), MemPlaceMeta::None, layout))
+    Ok(ecx.ptr_to_mplace(Pointer::from(alloc_id).into(), layout))
```
Comment (compiler-errors) [inline compiler/rustc_const_eval/src/interpret/util.rs:106]: does this change have anything to do w the pr? it seems unrelated lol

Comment (compiler-errors) [reply compiler/rustc_const_eval/src/interpret/util.rs:106]: oh, i see you changed the privacy of that fn.

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/util.rs:106]: Yeah, I had to add `ptr_to_mplace_unaligned` for this PR and so I had to make the lower-level `ptr_with_meta_to_mplace` more flexible and didn't want to expose this somewhat "dangerous" operation too far so I checked if I could make it private, and turns out yes I could.

**Review by compiler-errors — APPROVED** [review]
r=me unless you want specifically a review from saethlin

**Comment by RalfJung** [comment]
I had to rebase as there was a conflict (a semantic conflict -- things merged fine but then they would fail to build).

Seems fine to take this off saethlin's plate.
@bors r=compiler-errors
